### PR TITLE
New buttons implementation

### PR DESCRIPTION
**Pull request by fspoettel**
- [x] Split buttons into clean components
- [x] Add stories for new components
- [x] Add back stories for menu buttons & buttongroups
- [x] Add specs
- [x] Update dependent components

### CODE REVIEW

**Comment by philippfrank** [comment]
@fspoettel These are very good enhancements. I'd like to throw something in here:

This is the current syntax for a simple button with an icon. 

```jsx
<Button flavor="neutral" onClick={action()}>
  <ButtonIcon
    position="left"
    sprite="utility"
    icon="download"
  />
  Download
</Button>
```

I would like to write less code (with the option to still write exactly the above if needed). What about if we'd (optionally) offer something like:

```jsx
// default flavor
// default icon position
<Button onClick={action()} icon="utility.download">
  Download
</Button>
```

**Comment by fspoettel** [comment]
@philippfrank very good idea! Now that we have a dedicated `IconButton` and I solved the one case where this was not possible in the past (`more` prop for IconGroups), it is safe to add this. I'm a bit on the fence whether we should go with `icon="" sprite=""` or `icon="sprite.icon"` as you proposed since it would be the only place in the whole project where that is possible. We could implement `<IconButton icon="x" sprite="y" />` first as a shortcut and later implement your proposed syntax across all components. What do you think?

**Comment by philippfrank** [comment]
@fspoettel I am very open to anything which requires less code ;-)

**Comment by fspoettel** [comment]
@philippfrank the minimal signatures now are:

``` jsx
/* Same content/title */
<Button title="Bleep" />

/* Differing content/title */
<Button title="Bleep">Bloop</Button>

/*  Same content/title and left icon */
<Button title="Bleep" icon="foo" sprite="bar" />

/*  Differing content/title and left icon */
<Button title="Bleep" icon="foo" sprite="bar">Bloop</Button>

/* Right Icon */
<Button title="Bleep">
  <ButtonIcon
    position="right"
    icon="foo"
    sprite="bar"
  />Bleep
</Button>
/* Icon-Only button */
<IconButton icon="foo" sprite="bar" />

/* Stateful Icon-Only button */
<StatefulIconButton selected icon="foo" sprite="bar" />
```

The obvious worst-case is a button with a right-aligned icon. We could add a shortcut prop `position` and default it to `left` not sure whether that is worth it though. cc @brunottonurb

File: src/components/Button/Base/Button.js [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+      )}
+      {!isRightShortcut ? content : null}
+    </ButtonEl>
+  );
+};
+
+Button.defaultProps = {
+  children: null,
+  className: null,
+  flavor: 'neutral',
```
Comment (philippfrank) [inline src/components/Button/Base/Button.js]: 👍

File: src/components/Button/Base/Button.js [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+  href: null,
+  icon: null,
+  iconPosition: null,
+  sprite: null,
+  title: null,
+};
+
+Button.propTypes = {
+  /**
+   * Used to set content. Can be a string, a `ButtonIcon` or a combination of those
```
Comment (philippfrank) [inline src/components/Button/Base/Button.js]: Technically it `can` be everything. I'd remove that addendum -- it implies specific child types.

File: src/components/Button/Base/Button.js [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+   */
+  flavor: PropTypes.oneOf([
+    'neutral',
+    'brand',
+    'destructive',
+    'success',
+    'inverse'
+  ]),
+  /**
+   * Optional href, renders as `a` when set.
```
Comment (philippfrank) [inline src/components/Button/Base/Button.js]: Remove interpunction

File: src/components/Button/Base/StatefulButtonState.js [hunk 4]
```diff
@@ -0,0 +1,66 @@
...
+    state,
+    title,
+    ...rest
+  } = props;
+
+  /* eslint-disable no-nested-ternary */
+  const selectedClass = state === 'selected'
+    ? 'slds-text-selected'
+    : state === 'focus' ? 'slds-text-selected-focus' : 'slds-text-not-selected';
+  /* eslint-enable */
```
Comment (philippfrank) [inline src/components/Button/Base/StatefulButtonState.js]:
```es6
let selectedClass = null;
switch (state) {
  case 'selected':
    // ...
  case 'focus':
    // ...
  default:
    // ... 
}
```

**Comment by philippfrank** [comment]
Awesome enhancements @fspoettel 💯

**Pull request merged**